CLOUDSTACK-9480,  CLOUDSTACK-9495 fix egress rule incorrect behavior

murali-reddy opened the pull request:
When 'default egress policy' is set to 'allow' in the network offering, any egress rule that is added will 'deny' the traffic overriding the default behaviour.

Conversely, when 'default egress policy' is set to 'deny' in the network offering, any egress rule that is added will 'allow' the traffic overriding the default behaviour.

While this works for 'tcp', 'udp' as expected, for 'icmp' protocol its always set to ALLOW. This patch keeps all protocols behaviour consistent.

Results of running test/integration/component/test_egress_fw_rules.py.  With out the patch test_02_egress_fr2 test was failing. This patch fixes the test_02_egress_fr2  scenario. 
-----------------------------------------------------------------------------------------------------
Test By-default the communication from guest n/w to public n/w is NOT allowed. ... === TestName: test_01_1_egress_fr1 | Status : SUCCESS ===
ok
Test By-default the communication from guest n/w to public n/w is allowed. ... === TestName: test_01_egress_fr1 | Status : SUCCESS ===
ok
Test Allow Communication using Egress rule with CIDR + Port Range + Protocol. ... === TestName: test_02_1_egress_fr2 | Status : SUCCESS ===
ok
Test Allow Communication using Egress rule with CIDR + Port Range + Protocol. ... === TestName: test_02_egress_fr2 | Status : SUCCESS ===
ok
Test Communication blocked with network that is other than specified ... === TestName: test_03_1_egress_fr3 | Status : SUCCESS ===
ok
Test Communication blocked with network that is other than specified ... === TestName: test_03_egress_fr3 | Status : SUCCESS ===
ok
Test Create Egress rule and check the Firewall_Rules DB table ... === TestName: test_04_1_egress_fr4 | Status : SUCCESS ===
ok
Test Create Egress rule and check the Firewall_Rules DB table ... === TestName: test_04_egress_fr4 | Status : SUCCESS ===
ok
Test Create Egress rule and check the IP tables ... SKIP: Skip
Test Create Egress rule and check the IP tables ... SKIP: Skip
Test Create Egress rule without CIDR ... === TestName: test_06_1_egress_fr6 | Status : SUCCESS ===
ok
Test Create Egress rule without CIDR ... === TestName: test_06_egress_fr6 | Status : SUCCESS ===
ok
Test Create Egress rule without End Port ... === TestName: test_07_1_egress_fr7 | Status : EXCEPTION ===
ERROR
Test Create Egress rule without End Port ... === TestName: test_07_egress_fr7 | Status : SUCCESS ===
ok
Test Port Forwarding and Egress Conflict ... SKIP: Skip
Test Port Forwarding and Egress Conflict ... SKIP: Skip
Test Delete Egress rule ... === TestName: test_09_1_egress_fr9 | Status : SUCCESS ===
ok
Test Delete Egress rule ... === TestName: test_09_egress_fr9 | Status : SUCCESS ===
ok
Test Invalid CIDR and Invalid Port ranges ... === TestName: test_10_1_egress_fr10 | Status : SUCCESS ===
ok
Test Invalid CIDR and Invalid Port ranges ... === TestName: test_10_egress_fr10 | Status : SUCCESS ===
ok
Test Regression on Firewall + PF + LB + SNAT ... === TestName: test_11_1_egress_fr11 | Status : SUCCESS ===
ok
Test Regression on Firewall + PF + LB + SNAT ... === TestName: test_11_egress_fr11 | Status : SUCCESS ===
ok
Test Reboot Router ... === TestName: test_12_1_egress_fr12 | Status : SUCCESS ===
ok
Test Reboot Router ... === TestName: test_12_egress_fr12 | Status : EXCEPTION ===
ERROR
Test Redundant Router : Master failover ... === TestName: test_13_1_egress_fr13 | Status : SUCCESS ===
ok
Test Redundant Router : Master failover ... === TestName: test_13_egress_fr13 | Status : SUCCESS ===
ok
-----------------------------------------------------------------------------------------------------

murali-reddy commented:
@JayapalUradi could you please review

murali-reddy commented:
Test test_07_1_egress_fr7 failure was VM deploy failure. Re-run this particular test again, to enusre it succeeds.

rhtyd commented:
LGTM. @jburwell we should have this in 4.9.1.0-RC1.

murali-reddy commented:
@abhinandanprateek @borisstoyanov review?

borisstoyanov commented:
LGTM based on code review.

jayapalu commented on systemvm/patches/debian/config/opt/cloud/bin/configure.py:
```diff
@@ -145,40 +145,41 @@ def add_rule(self, cidr):
+                fwr = " -I FW_EGRESS_RULES"
+
+                #In case we have a default rule (accept all or drop all), we have to evaluate the action again.
+                if self.rule['cidr'] == ['0.0.0.0/0']:
```
Here the default rule when egress policy is true is identified using the cidr '0.0.0.0/0'. What if we configure a egress rule  (after the network is up ) with cidr '0.0.0.0/0' (ex: cidr 0.0.0.0/0, tcp, 22 egress policy on network:true)

murali-reddy replied:
If-else block does exactly that logic. If rule has CIDR 0.0.0.0/0 then rule action is set to that of default egress policy in the network offering. If rule has a CIDR other than 0.0.0.0/0 then reverse action (as that of default egress policy in the network offering) is set for the rule

jburwell commented:
@rhtyd let's hold this fix for 4.9.2.0.  I want to make sure we have had the opportunity to thoroughly test it before merge.

jburwell commented:
@murali-reddy could you please either add or update a Marvin test case to verify this fix?  Also, please squash your commits and add a JIRA ticket reference.  Finally, could you please investigate the Jenkins failure?

karuturi commented:
Started our internal CI to run smoke tests on this PR. It will post the results here. Thanks.

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 93
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=1
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failing since 11 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_ssvm.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

jburwell commented:
@murali-reddy could you please investigate the cause of the Jenkins build failure?

karuturi commented on systemvm/patches/debian/config/opt/cloud/bin/configure.py:
```diff
@@ -145,40 +145,41 @@ def add_rule(self, cidr):
+
                 if rule['protocol'] == "icmp":
                     self.fw.append(["filter", "front",
                                     " -A FW_EGRESS_RULES" +
```
I think this should be -I FW_EGRESS_RULES.

karuturi commented on systemvm/patches/debian/config/opt/cloud/bin/configure.py:
```diff
@@ -145,40 +145,41 @@ def add_rule(self, cidr):
+                           " -m %s " % rule['protocol'] + \
+                           " --dport %s" % rnge
+
+                self.fw.append(["filter", "", "%s -j %s" % (fwr, rule['action'])])
```
Incase of icmp, there will be two iptable rules. Its calling self.fw.append once in if(protocol)==icmp and another time outside. 
This line should be part of if(protocol != all and protocol != icmp)

karuturi commented on systemvm/patches/debian/config/opt/cloud/bin/configure.py:
```diff
@@ -145,40 +145,41 @@ def add_rule(self, cidr):
-
-                    self.fw.append(["filter", "", "%s -j %s" % (fwr, rule['action'])])
+
+                if rule['protocol'] != "all":
```
This if should be if(protocol != all and protocol != icmp) to avoid executing again incase of icmp

murali-reddy commented:
@jburwell i have squashed the commit. Added two new test cased to test /32 CIDR for egress firewall rule. For some reason jenkins seems to have failed due to unrelated unit test failure. I will try to trigger new build

murali-reddy commented:
@karuturi i did not change the logic of how and what iptables rules were added in the original patch. I just stream lined rule action deny/allow logic for all the protocols. 

I have updated the patch as there is one more egress rule bug "CLOUDSTACK-9495: Egress rules functionalty broken when protocol=all specified". Now iptable rules are only inserted in the front.

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 100
 Hypervisor xenserver
 NetworkType Advanced
 Passed=76
 Failed=15
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_vm_snapshots.py

 * test_01_create_vm_snapshots Failed

 * test_02_revert_vm_snapshots Failed

 * test_03_delete_vm_snapshots Failed

* test_service_offerings.py

 * ContextSuite context=TestServiceOfferings>:setup Failing since 13 runs

* test_loadbalance.py

 * test_01_create_lb_rule_src_nat Failed

 * test_02_create_lb_rule_non_nat Failed

 * test_assign_and_removal_lb Failed

* test_router_dns.py

 * test_router_dns_guestipquery Failed

* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failed

* test_volumes.py

 * test_01_create_volume Failed

 * test_02_attach_volume Failed

* test_vm_life_cycle.py

 * test_08_migrate_vm Failed

 * test_10_attachAndDetach_iso Failed

* test_routers_network_ops.py

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failed

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_ssvm.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

jburwell commented:
@murali-reddy have you had a chance to investigate the Jenkins failure(s) and the test failures identified by @karuturi?  

Does this PR address [CLOUDSTACK-9480](https://issues.apache.org/jira/browse/CLOUDSTACK-9480)?  If so, please amend the commit message to reflect this association?  Otherwise, which JIRA ticket should be associated with this issue?

In terms of release target, this issue seems to affect 4.8 as well?  How much effort would it be to re-target it to 4.8 for pending 4.8.2.0 release?

murali-reddy commented:
@jburwell Jenkins was failing unrelated unit-test failure. I have triggered new Jenkins build which has passed. Changed the title of the PR to reflect that its fixing issues CLOUDSTACK-9480, CLOUDSTACK-9495. Will send a new PR for 4.8.

@karuturi @cloudmonger failed tests seems to have failed in build #99 as well. Is it environment issue? Failure in test_vm_snapshots.py are totally unrelated to changes in the fix.

karuturi commented:
@murali-reddy recurring failures will be listed as failing since last n runs. 
The router related tests(lb, network ops) are failing due to this PR. 
Are the tests running(manual or BVT) at your end after the new changes?

karuturi commented:
I think in the latest code change, `fwr` is not defined before and hence `fwr +=` operation will fail.
Also, the ICMP rule is still appended and not Inserted in the beginning. Is that ok? (I dont have enough iptables knowledge.)

murali-reddy commented:
@karuturi thanks for pointing the issue with fwr. i messed up patch while squashing commits. I fixed the patch, waiting to trigger internal smoke test run. Will update the PR once i have test results

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-16

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 101
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=1
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failing since 2 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_ssvm.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

murali-reddy commented:
@cloudmonger thanks

test_extendPhysicalNetworkVlan failed due to 'physicalnetwork 200 has allocated vnets in the range 734-735'. Possibly due to stale network that was not cleaned by one of the tests.

Ran test manually, it passed.
